lib/clerk/authenticate_request.rb

```diff
@@ -0,0 +1,245 @@
+            if auth_context.handshake_token?
+                return resolve_handshake(env)
+            end
+
+            if auth_context.development_instance? && auth_context.dev_browser?
+                return handle_handshake_maybe_status(env, reason: AuthErrorReason::DEV_BROWSER_SYNC)
+            end
+
+            # TODO(dimkl): Add multi-domain support for production
+            # if auth_context.production_instance? && auth_context.eligible_for_multi_domain?
+                # return handle_handshake_maybe_status(env, reason: AuthErrorReason::SATELLITE_COOKIE_NEEDS_SYNCING)
+            # end
+
+            # TODO(dimkl): Add multi-domain support for development
+            # if auth_context.development_instance? && auth_context.eligible_for_multi_domain?
+                # trigger handshake using auth_context.sign_in_url as base redirect_url
+                # return handle_handshake_maybe_status(env, reason: AuthErrorReason::SATELLITE_COOKIE_NEEDS_SYNCING, '', headers);
+            # end
+
+            # TODO(dimkl): Add multi-domain support for development in primary
+            # if auth_context.development_instance? && !auth_context.is_satellite? && auth_context.clerk_redirect_url
+                # trigger handshake using auth_context.clerk_redirect_url as base redirect_url + mark it as clerk_synced
+                # return handle_handshake_maybe_status(env, reason: AuthErrorReason::PRIMARY_RESPONDS_TO_SYNCING, '', headers);
+            # end
+
+            if !auth_context.active_client? && !auth_context.session_token_in_cookie?
+                return signed_out(reason: AuthErrorReason::SESSION_TOKEN_AND_UAT_MISSING)
+            end
+
+            # This can eagerly run handshake since client_uat is SameSite=Strict in dev
+            if !auth_context.active_client? && auth_context.session_token_in_cookie?
+                return handle_handshake_maybe_status(env, reason: AuthErrorReason::SESSION_TOKEN_WITHOUT_CLIENT_UAT)
+            end
+
+            if auth_context.active_client? && !auth_context.session_token_in_cookie?
+                return handle_handshake_maybe_status(env, reason: AuthErrorReason::CLIENT_UAT_WITHOUT_SESSION_TOKEN)
+            end
+
+            begin
+                token = verify_token(auth_context.session_token_in_cookie)
+                return signed_out if !token
+
+                if token["iat"] < auth_context.client_uat
+                    return handle_handshake_maybe_status(env, reason: AuthErrorReason::SESSION_TOKEN_OUTDATED);
+                end
+
+                return signed_in(env, token, auth_context.session_token_in_cookie)
+            rescue JWT::ExpiredSignature
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_EXPIRED)
+            rescue JWT::InvalidIatError
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_NOT_ACTIVE_YET)
+            end
+
+            signed_out
+        end
+
+        def resolve_handshake(env)
```

Comment:
:cry: This method is a bit hard to follow. Might be just me.

Reply:
Would comments help? The main idea here is that if you find the `__handshake_token` query parameter, then verify/decode the value (It's a JWT) and assign the cookie directives find in the decoded value as headers in the response. If there is no `__session` directive then it's signed-out request, else you verify the session token inside the `__session` directive. For development instances if the session token inside the handshake token is expired or not active yet, we try to verify it with a big leeway (1 day) to skip clock skew issues. If all of those fail, we raise an error because something does not work as expected in the handshake mechanism.

I don't know if the original implementation is easier to follow but this is the original implementation in JS i ported to Ruby: https://github.com/clerk/javascript/blob/8e5c881c40d7306c5dbd2e1f1803fbf75127bd71/packages/backend/src/tokens/request.ts#L92

Reply:
Thanks for the explanation! Comments do help, and I also think that extracting methods, renaming fields etc. will help.

This is definitely something that can be done at a later stage, don't let it delay your work.

Reply:
i will add some comments and try to split this method to smaller building blocks to improve readability.